README.md
# koa-yup-router-docs

Generates [OpenApi 3 document](https://swagger.io/specification/#openapi-object) with paths from [koa yup router][koa-yup-router].

[![NPM version][npm-image]][npm-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/@rudi23/koa-yup-router-docs.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@rudi23/koa-yup-router-docs
[download-image]: https://img.shields.io/npm/dm/@rudi23/koa-yup-router-docs.svg?style=flat-square
[download-url]: https://www.npmjs.com/package/@rudi23/koa-yup-router-docs
[koa-yup-router]: https://github.com/rudi23/koa-yup-router

### Installation

Install using [`npm`][npm-url]:

```bash
npm install @rudi23/koa-yup-router-docs
```

NodeJS `>= 12.0.0.` is required.

### Example

```js
import * as yup from 'yup';
import YupRouter from '@rudi23/koa-yup-router';
import createDocument from '@rudi23/koa-yup-router-docs';

const router = new YupRouter();

router.addRoute({
    method: 'get',
    path: '/user',
    handler: () => {},
});

router.addRoute({
    method: ['post', 'put'],
    path: '/user/:id',
    validate: {
        type: 'json',
        body: yup.object({
            firstName: yup.string().required(),
            lastName: yup.string().required(),
            age: yup.number().required().min(0).max(100),
        }),
        params: yup.object({
            id: yup.number().required(),
        }),
        headers: yup.object({
            custom: yup.string(),
        }),
        output: {
            200: {
                description: 'Success OK',
                body: yup.string().required(),
            },
            201: {
                description: 'Created OK',
                body: {
                    'application/json': yup.object({
                        id: yup.number().required(),
                    }),
                },
            },
            400: {
                description: 'Bad request',
                headers: {
                    'X-Rate-Limit-Limit': {
                        schema: yup.number().required(),
                        description: 'The number of allowed requests in the current period',
                    },
                },
            },
        },
    },
    handler: () => {},
});

router.addRoute({
    method: 'delete',
    path: '/user/:id',
    validate: {
        type: 'json',
        params: yup.object({
            id: yup.number().required(),
        }),
        headers: yup.object({
            custom: yup.string(),
        }),
    },
    handler: () => {},
});

const openApiDoc = createDocument(router, {
    openapi: '3.0.3',
    info: {
        title: 'Sample API',
        description: 'Sample API',
        version: '1.0.0',
    },
    servers: [
        {
            url: 'https://www.example.com/api/v1',
        },
    ],
    paths: {},
    components: {
        securitySchemes: {
            BasicAuth: {
                type: 'apiKey',
                name: 'Authorization',
                in: 'header',
            },
        },
    },
});
```

## LICENSE

[MIT](https://github.com/rudi23/koa-yup-router-docs/blob/master/LICENSE)


CHANGELOG.md
## [2.2.6](https://github.com/rudi23/koa-yup-router-docs/compare/v2.2.5...v2.2.6) (2022-09-12)

### Bug Fixes

-   **deps:** update dependency @rudi23/koa-yup-router to v2.2.4 ([#68](https://github.com/rudi23/koa-yup-router-docs/issues/68)) ([f2cfa8f](https://github.com/rudi23/koa-yup-router-docs/commit/f2cfa8f9b9e097d492c4535bd9302368d1ee05e8))

### Miscellaneous Chores

-   **deps:** pin dependency @types/koa\_\_router to 8.0.11 ([bd509df](https://github.com/rudi23/koa-yup-router-docs/commit/bd509df11139d9b3d1fcd1048e037982591a8305))
-   **deps:** update dependency @commitlint/cli to v17.1.2 ([9893d5f](https://github.com/rudi23/koa-yup-router-docs/commit/9893d5f46a3492e724beb17e62bd79a28f43a3d5))
-   **deps:** update dependency @release-it/conventional-changelog to v5.1.0 ([d379108](https://github.com/rudi23/koa-yup-router-docs/commit/d37910822a4aa8620d48d5fd659d482fea16461d))
-   **deps:** update dependency @types/jest to v28.1.7 ([1ef8684](https://github.com/rudi23/koa-yup-router-docs/commit/1ef8684d08e66af348c75e2355b3b97cd3e611cf))
-   **deps:** update dependency @types/jest to v28.1.8 ([09a1c46](https://github.com/rudi23/koa-yup-router-docs/commit/09a1c4650618a1e90e02d23e0bf6cfc6689c8cf6))
-   **deps:** update dependency @types/koa\_\_router to v12 ([#65](https://github.com/rudi23/koa-yup-router-docs/issues/65)) ([132370b](https://github.com/rudi23/koa-yup-router-docs/commit/132370b74b5cb2ed7d670c1b5bac222ab2017999))
-   **deps:** update dependency eslint to v8.22.0 ([69e49a9](https://github.com/rudi23/koa-yup-router-docs/commit/69e49a9449f9bb2385e499048aa5ffe28be0bfbf))
-   **deps:** update dependency eslint to v8.23.0 ([8da6454](https://github.com/rudi23/koa-yup-router-docs/commit/8da645426e135e278943d193587c6448559c34e2))
-   **deps:** update dependency eslint to v8.23.1 ([c903f0b](https://github.com/rudi23/koa-yup-router-docs/commit/c903f0b3ec5052eb0cae2daedd8475272c84d480))
-   **deps:** update dependency eslint-import-resolver-typescript to v3.4.1 ([359d4c9](https://github.com/rudi23/koa-yup-router-docs/commit/359d4c98e5434ffffcc637e28e10ecf19d06f180))
-   **deps:** update dependency eslint-import-resolver-typescript to v3.4.2 ([0a7870b](https://github.com/rudi23/koa-yup-router-docs/commit/0a7870b47001587e8a2ca1e850d81ecaea6fad37))
-   **deps:** update dependency eslint-import-resolver-typescript to v3.5.1 ([e23aeb7](https://github.com/rudi23/koa-yup-router-docs/commit/e23aeb7e647b67787a82c014e20835cb6e0fb332))
-   **deps:** update dependency eslint-plugin-jest to v26.7.0 ([43c960d](https://github.com/rudi23/koa-yup-router-docs/commit/43c960dbe9789d1588f0cc26cab5b0efcaa241ff))
-   **deps:** update dependency eslint-plugin-jest to v26.8.0 ([6344e0f](https://github.com/rudi23/koa-yup-router-docs/commit/6344e0fc99d3b11e977682f0a79c1762dd6381c4))
-   **deps:** update dependency eslint-plugin-jest to v26.8.2 ([ab0100b](https://github.com/rudi23/koa-yup-router-docs/commit/ab0100b9d8dbbf5fe97e4bbb40b30f20ce201797))
-   **deps:** update dependency eslint-plugin-jest to v26.8.3 ([2666869](https://github.com/rudi23/koa-yup-router-docs/commit/2666869a754a56268cc687dbdbbc5a9b19e7a6ae))
-   **deps:** update dependency eslint-plugin-jest to v26.8.4 ([3c4dced](https://github.com/rudi23/koa-yup-router-docs/commit/3c4dcede0baf78707e97e3235003e96d09bf6bad))
-   **deps:** update dependency eslint-plugin-jest to v26.8.5 ([93b3be1](https://github.com/rudi23/koa-yup-router-docs/commit/93b3be1aefdafd27fcc41bbf257528eb33e4effe))
-   **deps:** update dependency eslint-plugin-jest to v26.8.7 ([8e2e9ad](https://github.com/rudi23/koa-yup-router-docs/commit/8e2e9ad64b832573003e976a3a900569ae3e721e))
-   **deps:** update dependency eslint-plugin-jest to v26.9.0 ([dbc8b43](https://github.com/rudi23/koa-yup-router-docs/commit/dbc8b431052775e3976ec18e15c9400ab3dad8a4))
-   **deps:** update dependency eslint-plugin-jest to v27 ([#57](https://github.com/rudi23/koa-yup-router-docs/issues/57)) ([315e6d4](https://github.com/rudi23/koa-yup-router-docs/commit/315e6d444b8b2d96fc632dd34b7d964ba6fb3b5d))
-   **deps:** update dependency release-it to v15.3.0 ([a15b9db](https://github.com/rudi23/koa-yup-router-docs/commit/a15b9db63d25460c8a036b8159579275d728f255))
-   **deps:** update dependency release-it to v15.4.0 ([7f00e91](https://github.com/rudi23/koa-yup-router-docs/commit/7f00e91b1ca31ed276f16f1d3a704568f8cf4b39))
-   **deps:** update dependency release-it to v15.4.1 ([07fc3d4](https://github.com/rudi23/koa-yup-router-docs/commit/07fc3d4b3bd802ffda60854eeebefbc5eb7c31c6))
-   **deps:** update dependency ts-jest to v28.0.8 ([60fa09e](https://github.com/rudi23/koa-yup-router-docs/commit/60fa09ee469c435b02dad58164315c6800615e47))
-   **deps:** update dependency typescript to v4.8.2 ([fc9b2f6](https://github.com/rudi23/koa-yup-router-docs/commit/fc9b2f6c5ba07caaae12047b928c30fbaee4878b))
-   **deps:** update dependency typescript to v4.8.3 ([b910c73](https://github.com/rudi23/koa-yup-router-docs/commit/b910c73c84a0851b7a85121c165816cd77b2cd10))
-   **deps:** update linters ([fc9e87e](https://github.com/rudi23/koa-yup-router-docs/commit/fc9e87e45cd15553531e0969ca9bdfdd4a8fe720))
-   **deps:** update linters ([43a009e](https://github.com/rudi23/koa-yup-router-docs/commit/43a009ea169f2bb28e642d1a3f58eb139a3b0e3b))
-   **deps:** update linters ([9ae49db](https://github.com/rudi23/koa-yup-router-docs/commit/9ae49dba3c3ff3a0cbb77be3bd405e2729d06ef9))
-   **deps:** update linters to v5.33.1 ([ba3a3e1](https://github.com/rudi23/koa-yup-router-docs/commit/ba3a3e1fafae1e45bd0706a97989e57951b249dc))
-   **deps:** update linters to v5.35.1 ([7ec3eb9](https://github.com/rudi23/koa-yup-router-docs/commit/7ec3eb975f3be45b3770f37d3d137fca79acdc85))
-   **deps:** update linters to v5.36.0 ([8544e5a](https://github.com/rudi23/koa-yup-router-docs/commit/8544e5a3ce0fa59b253dd681a7c5be673ce7d037))

src/index.ts
import merge from 'lodash.merge';
import { isSchemaObject } from 'openapi3-ts';
import { isSchema } from 'yup';
import yupToOpenapi from '@rudi23/yup-to-openapi';
import type YupRouter from '@rudi23/koa-yup-router';
import type { RouteSpecification, InputType, ValidationSchema } from '@rudi23/koa-yup-router';
import type {
    OperationObject,
    ParameterLocation,
    ParameterObject,
    PathsObject,
    OpenAPIObject,
    ContentObject,
    HeadersObject,
    ResponseObject,
    ResponsesObject,
} from 'openapi3-ts';
import type { OutputSchema } from '@rudi23/koa-yup-router/@type/types/index.js';
import type { Config } from './types/index.js';

const defaultConfig: Config = {
    headRoutes: false,
    defaultResponses: {
        200: {
            description: 'OK',
        },
        400: {
            description: 'Bad request',
        },
        500: {
            description: 'Internal Server Error',
        },
    },
};

function getParameters(yupSchema: ValidationSchema, inPath: ParameterLocation): ParameterObject[] {
    const schemaObj = yupToOpenapi(yupSchema);

    return Object.entries(schemaObj.properties || {}).map(([name, props]) => {
        if (isSchemaObject(props)) {
            const { description, ...restProps } = props;

            const fieldSchema: ParameterObject = {
                name,
                in: inPath,
                schema: restProps,
                required: schemaObj.required?.includes(name),
            };
            if (description) {
                fieldSchema.description = description;
            }

            return fieldSchema;
        }

        return {
            name,
            in: inPath,
            schema: {
                $ref: props.$ref,
            },
        };
    });
}

function getMediaType(input: InputType | undefined): string {
    switch (input) {
        case 'json':
            return 'application/json';
        case 'form':
            return 'application/x-www-form-urlencoded';
        case 'multipart':
            return 'multipart/form-data';
        default:
            return '*/*';
    }
}

function getResponseContent(outputBodySchema: OutputSchema['body']): ContentObject | undefined {
    if (!outputBodySchema || Object.keys(outputBodySchema).length === 0) {
        return undefined;
    }

    if (isSchema(outputBodySchema)) {
        return {
            'application/json': {
                schema: yupToOpenapi(outputBodySchema),
            },
        };
    }

    return Object.entries(outputBodySchema).reduce((aggSchema, [contentType, schema]) => {
        if (isSchema(schema)) {
            return {
                ...aggSchema,
                [contentType]: { schema: yupToOpenapi(schema) },
            };
        }

        return aggSchema;
    }, {} as ContentObject);
}

function getResponseHeaders(outputHeadersSchema: OutputSchema['headers']): HeadersObject | undefined {
    if (!outputHeadersSchema || Object.keys(outputHeadersSchema).length === 0) {
        return undefined;
    }

    return Object.entries(outputHeadersSchema).reduce((aggSchema, [contentType, schema]) => {
        if (isSchema(schema.schema)) {
            return {
                ...aggSchema,
                [contentType]: { schema: yupToOpenapi(schema.schema), description: schema.description },
            };
        }

        return aggSchema;
    }, {} as HeadersObject);
}

function getResponses(specs: RouteSpecification, config: Config): ResponsesObject {
    if (specs.meta?.swagger?.response) {
        return specs.meta.swagger.response as ResponsesObject;
    }

    const responsesSpec = merge({}, config.defaultResponses, specs.validate.output);
    if (responsesSpec) {
        const responses = {} as ResponsesObject;
        Object.entries(responsesSpec).forEach(([code, responseSpec]) => {
            if (!responseSpec || Object.keys(responseSpec).length === 0) {
                return;
            }

            const { body, headers, description } = responseSpec;

            const response = {} as ResponseObject;
            const responseDescription = description;
            if (responseDescription) {
                response.description = responseDescription;
            }
            const responseContent = getResponseContent(body);
            if (responseContent) {
                response.content = responseContent;
            }
            const responseHeaders = getResponseHeaders(headers);
            if (responseHeaders) {
                response.headers = responseHeaders;
            }

            responses[code] = response;
        });

        return responses;
    }

    return {};
}

function createOperationObject(specs: RouteSpecification, config: Config): OperationObject {
    const schema: OperationObject = {
        responses: {},
    };

    schema.responses = getResponses(specs, config);

    let parameters: ParameterObject[] = [];
    if (specs.validate.params) {
        parameters = [...parameters, ...getParameters(specs.validate.params, 'path')];
    }
    if (specs.validate.query) {
        parameters = [...parameters, ...getParameters(specs.validate.query, 'query')];
    }
    if (specs.validate.headers) {
        parameters = [...parameters, ...getParameters(specs.validate.headers, 'header')];
    }
    if (parameters.length > 0) {
        schema.parameters = parameters;
    }

    if (specs.meta?.swagger?.summary) {
        schema.summary = specs.meta?.swagger?.summary;
    }
    if (specs.meta?.swagger?.description) {
        schema.description = specs.meta?.swagger?.description;
    }
    if (specs.meta?.swagger?.deprecated) {
        schema.deprecated = specs.meta?.swagger?.deprecated;
    }
    if (specs.meta?.swagger?.tags) {
        schema.tags = specs.meta?.swagger?.tags;
    }
    if (specs.meta?.swagger?.security) {
        schema.security = specs.meta?.swagger?.security;
    }
    if (specs.validate.body) {
        const mediaType = getMediaType(specs.validate.type);
        schema.requestBody = {
            content: {
                [mediaType]: {
                    schema: yupToOpenapi(specs.validate.body),
                },
            },
        };
    }

    return schema;
}

function processRoute(route: YupRouter, config: Config): PathsObject {
    return route.routeSpecs.reduce((aggPs, specs, index) => {
        const layer = route.router.stack[index];
        const path = layer.path.replace(/:(\w+)/g, '{$1}');
        const schema = createOperationObject(specs, config);

        layer.methods
            .map((m) => m.toLowerCase())
            .filter((m) => !config.headRoutes && m !== 'head')
            .forEach((method) => {
                // eslint-disable-next-line no-param-reassign
                aggPs[path] = {
                    ...aggPs[path],
                    [method]: schema,
                };
            });

        return aggPs;
    }, {} as PathsObject);
}

function createPaths(router: YupRouter | YupRouter[], config: Partial<Config> = {}): PathsObject {
    const mergedConfig = { ...defaultConfig, ...config };
    const routes = !Array.isArray(router) ? [router] : router;

    return routes.reduce(
        (aggRoutes, route) => ({
            ...aggRoutes,
            ...processRoute(route, mergedConfig),
        }),
        {} as PathsObject
    );
}

function createDocument(
    router: YupRouter | YupRouter[],
    initialDoc: OpenAPIObject,
    config: Partial<Config> = {}
): OpenAPIObject {
    const paths = createPaths(router, config);

    return {
        ...initialDoc,
        paths: { ...initialDoc.paths, ...paths },
    };
}

export default createDocument;
export { createDocument, createPaths };
export * from './types/index.js';
export * from 'openapi3-ts';
